Mathlib/RingTheory/Polynomial/Eisenstein/Generalized.lean

```diff
@@ -0,0 +1,232 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Algebra.Field.ZMod
+import Mathlib.Algebra.Polynomial.FieldDivision
+import Mathlib.Algebra.Polynomial.SpecificDegree
+import Mathlib.Data.ZMod.QuotientRing
+import Mathlib.Tactic.ComputeDegree
+
+/-! # A generalized Eisenstein criterion
+
+`Polynomial.generalizedEisenstein` :
+  Let `R` be an integral domain and `K` an `R`-algebra which is a field.
+  Let `q : R[X]` be a monic polynomial which is irreducible in `K[X]`.
+  Let `f : R[X]` be a monic polynomial of strictly positive degree
+  whose image in `K[X]` is a power of `q`.
+  Assume moreover that `f.modByMonic q` is not zero in `(R ⧸ (P ^ 2))[X]`,
+  where `P` is the kernel of `algebraMap R K`.
+  Then `f` is irreducible.
+
+The Eisenstein criterion is the particular case where `q := X`.
+
+The case of a polynomial `q := X - a` is interesting,
+then the mod `P ^ 2` hypothesis can rephrased as saying
+that `f.derivative.eval a ∉ P ^ 2`. (TODO)
+The case of cyclotomic polynomials of prime index `p` could be proved directly
+using that result, taking `a = 1`; the derivative is `p`.
+
+We give a (possibly non convincing) application to the irreducibility
+of the polynomial `X ^ 4 - 10 * X + 1` in `ℤ[X]`.
+One argues modulo `3`, with `q := X ^ 2 + 1`.
+
+## Remark
+
+The result can also be generalized to the case where
+the leading coefficients of `f` and `q` do not belong to `P`.
+(By localization at `P`, make these coefficients invertible.)
+There are two obstructions, though :
+
+* Usually, one will only obtain irreducibility in `F[X]`, where `F` is the field
+of fractions of `R`. (If `R` is a UFD, this will be close to what is wanted,
+but not in general.)
+
+* The mod `P ^ 2` hypothesis will have to be rephrased to a condition
+in the second symbolic power of `P`. When `P` is a maximal ideal,
+that symbolic power coincides with `P ^ 2`, but not in general.
+
+-/
+
+open Polynomial Ideal.Quotient
+
+namespace Polynomial
+
+variable {R : Type*} [CommRing R] {K : Type*} [Field K] [Algebra R K]
+
+theorem exists_C_leadingCoeff_mul_pow_of_dvd_pow
```

Comment:
You could weaken the assumptions just requiring that `K` be a domain and `hq` be prime, no?

Reply:
Indeed, I have even done so in the main theorem.